include/fi_mem.h

```diff
@@ -69,24 +69,31 @@ static inline void *mem_dup(const void *src, size_t size)
  */
 #define FREESTACK_EMPTY	NULL
 
+#define FREESTACK_HEADER 					\
+	size_t		size;					\
+	void		*next;					\
+
 #define freestack_isempty(fs)	((fs)->next == FREESTACK_EMPTY)
 #define freestack_push(fs, p)					\
 {								\
 	*(void **) p = (fs)->next;				\
 	(fs)->next = p;						\
 }
-#define freestack_pop(fs)					\
-	({							\
-		void *p = (fs)->next;				\
-		assert(!freestack_isempty(fs));			\
-		(fs)->next = *((void **) p);			\
-		p;						\
-	})
+#define freestack_pop(fs) freestack_pop_impl(fs, fs->next)
+
+static inline void* freestack_pop_impl(void *fs, void *fs_next)
+{
+	struct {
+		FREESTACK_HEADER;
+	} *_fs = fs;
+	assert(!freestack_isempty(_fs));
+	(_fs)->next = *((void **)fs_next);
+	return fs_next;
+}
```

Comment:
My original idea was completely replace ```#define freestack_pop(fs)``` by inline function. But as you can see there is building of stack structure using macro and such structures have no hierarchy. That is why inline function should have ```void*``` pointer to stack object (and later cast it into stack). But... if we look at ```freestack_pop``` macro usage we will see code like: ```freestack_pop(cmap->keypool)```. Stack element is part of more global structure and if somebody write code like ```freestack_pop(cmap)``` then inline function will accept such pointer (compiler will not warn about it) and developer will spend many happy hours searching what is wrong